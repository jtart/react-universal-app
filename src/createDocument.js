import Helmet from 'react-helmet';

const createDocument = (appHTML, data, scripts, additionalHeadElements) => {
  const helmet = Helmet.renderStatic();
  const serialisedData = JSON.stringify(data).replace(/</g, '\\u003c');

  return `
    <!DOCTYPE html>
    <html>
      <head ${helmet.htmlAttributes.toString()}>
        ${helmet.title.toString()}
        ${helmet.meta.toString()}
        ${helmet.link.toString()}
        ${additionalHeadElements.join('')}
      </head>
      <body ${helmet.bodyAttributes.toString()}>
        <div id="__uni__">${appHTML}</div>
        <script>window.__UNI_DATA__ = ${serialisedData}</script>
        ${scripts
          .map(script => `<script src="${script}" defer></script>`)
          .join('')}
      </body>
    </html>
  `;
};

export default createDocument;
