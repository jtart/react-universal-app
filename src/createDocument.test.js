import createDocument from './createDocument';
import Helmet from 'react-helmet';

Helmet.renderStatic = jest.fn().mockReturnValue({
  htmlAttributes: jest.fn().mockReturnValue({ lang: 'en_GB' }),
  title: jest.fn().mockReturnValue('<title />'),
  meta: jest.fn().mockReturnValue('<meta />'),
  link: jest.fn().mockReturnValue(['<link key="l-1" />', '<link key="l-2" />']),
  bodyAttributes: jest.fn().mockReturnValue({ dir: 'ltr' }),
});

describe('createDocument', () => {
  it('should render correctly a string containing a HTML document', () => {
    const appHTML = '<h1>This is some html</h1>';
    const data = { data: 'This is some data!' };
    const scripts = [
      'https://example.com/one.js',
      'https://example.com/two.js',
    ];
    const additionalHeadElements = [
      '<style key="s-1" data-test-styles-one />',
      '<style key="s-2" data-test-styles-two />',
    ];
    const doc = createDocument(appHTML, data, scripts, additionalHeadElements);
    expect(doc).toMatchSnapshot();
  });
});
