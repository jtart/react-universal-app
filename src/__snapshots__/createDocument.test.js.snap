// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`createDocument should render correctly a string containing a HTML document 1`] = `
"
    <!DOCTYPE html>
    <html>
      <head function() {
        return fn.apply(this, arguments);
      }>
        function() {
        return fn.apply(this, arguments);
      }
        function() {
        return fn.apply(this, arguments);
      }
        function() {
        return fn.apply(this, arguments);
      }
        <style key=\\"s-1\\" data-test-styles-one /><style key=\\"s-2\\" data-test-styles-two />
      </head>
      <body function() {
        return fn.apply(this, arguments);
      }>
        <div id=\\"__uni__\\"><h1>This is some html</h1></div>
        <script>window.__UNI_DATA__ = {\\"data\\":\\"This is some data!\\"}</script>
        <script src=\\"https://example.com/one.js\\" defer></script><script src=\\"https://example.com/two.js\\" defer></script>
      </body>
    </html>
  "
`;
